java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java

```diff
@@ -385,10 +392,97 @@ private API getXPAPI(
     return API.UNKNOWN;
   }
 
-  private boolean isArgumentMatchesFlagName(ExpressionTree argTree, Symbol argSym) {
+  private boolean isArgumentMatchesFlagName(
+      ExpressionTree argTree, VisitorState state, Symbol argSym) {
     return (isLiteralTreeAndMatchesFlagName(argTree)
         || isVarSymbolAndMatchesFlagName(argSym)
-        || isSymbolAndMatchesFlagName(argSym));
+        || isSymbolAndMatchesFlagName(argSym)
+        || isMatchingEnumFieldValue(argTree, state));
+  }
+
+  /*
+   * Matches on an enum in an expression tree. Covers the following 4 cases:
+   * 1. Unqualified import enum constant (e.g. "STALE_FLAG")
+   * 2. Qualified enum name (e.g. "TestExperimentName.STALE_FLAG")
+   * 3. Unqualified enum constant with method call (e.g. "STALE_FLAG.getKey()")
+   * 4. Qualified enum name with method call (e.g. "TestExperimentName.STALE_FLAG.getKey()")
+   */
+  private boolean isMatchingEnumFieldValue(ExpressionTree argTree, VisitorState state) {
+    // Skip if no enum records are configured, to avoid excessive AST lookups
+    if (!this.config.hasEnumRecords()) {
+      return false;
+    }
+
+    Symbol sym;
+    if (argTree.getKind() == Kind.IDENTIFIER) {
+      // Case 1: Unqualified enum constant (e.g. "STALE_FLAG")
+      sym = ASTHelpers.getSymbol(argTree);
+    } else if (argTree.getKind() == Kind.MEMBER_SELECT) {
+      // Case 2: Qualified enum name (e.g. "TestExperimentName.STALE_FLAG")
+      MemberSelectTree memberTree = (MemberSelectTree) argTree;
+      sym = ASTHelpers.getSymbol(memberTree);
+    } else if (argTree.getKind() == Kind.METHOD_INVOCATION) {
+      MethodInvocationTree methodTree = (MethodInvocationTree) argTree;
+      if (methodTree.getMethodSelect().getKind() != Kind.MEMBER_SELECT) {
+        return false;
+      }
+      MemberSelectTree memberTree = (MemberSelectTree) methodTree.getMethodSelect();
+      if (memberTree.getExpression().getKind() == Kind.IDENTIFIER) {
+        // Case 3: Unqualified enum constant with method call (e.g. "STALE_FLAG.getKey()")
```

Comment:
Hmm so after thinking about this, the issue is if we remove the constant `STALE_FLAG`, it doesn't matter what method call we find, we need to remove all usages because we removed the enum constant itself.

There is a potential use case where the constant stays, and we remove ONLY based on method call. As of right now, that doesn't seem like a useful use case to me. (I could be wrong though!)

Will add a comment to clarify at the very least.

Reply:
Haven't done a full second pass yet, but that's a very valid point. Since the methods are instance methods, we also can't associate them just with the enum, so, removing the enum constant requires removing any calls to methods of that enum constant. Now, the question is, how do we simplify code around that call. Consider the following:

```
public enum ButtonColorFlag {
    // ... a hundred+ other flags
    BLUE_BUTTON_FLAG("button.is.blue", "#0000FF"),
    RED_BUTTON_FLAG("button.is.red", "#FF0000");

    private final String key;
    private final String value;

    public String getKey() {
        return key;
    }

    public String getValue() {
        return value;
    }

    FeatureFlags(final String key, final String value) {
        this.key = key;
        this.value = value;
    }
}
```

Now, presumably code calling `BLUE_BUTTON_FLAG.getKey()` will only happen inside whatever checks for whether the feature is toggled or not (`featureFlagClient.isFeatureEnabled` in #141), but `getValue()` could be called directly, with the value meaning something and, more importantly being assigned or passed to some other field/call:

```
this.button = new Button(color=BLUE_BUTTON_FLAG.getValue())
```

So, what is Piranha to do then?

a) Abort? (e.g. we just don't allow flag enums to carry values others than those needed to check the flag name/identifier)
b) Remove the line? Possibly leaving `this.button` unassigned and breaking code? (In our internal deployment, we don't require Piranha to always produce building code, as some refactorings will require manual follow up no matter what, but we do aim for it)
c) Try to propagate `value`? (Easy-ish in the example above, but impossible in the general case where the enum contains methods that aren't statically evaluatable)

Reply:
Good point. I actually just made a quick test case for it. You are correct, (b) is the current behavior, and this change preserves it, which makes sense to me. Wonder if we could in the future add some logic to Piranha to replace statically evaluatable methods of enum constant with their constant values but, in the limit, this is impossible to handle for all cases of enum methods.